tsdb/exemplar.go

```diff
@@ -0,0 +1,209 @@
+	exemplars []*circularBufferEntry
+	nextIndex int
+}
+
+type circularBufferEntry struct {
+	exemplar        exemplar.Exemplar
+	seriesLabels    labels.Labels
+	scrapeTimestamp int64
+	prev            int // index of previous exemplar in circular for the same series, use -1 as a default for new entries
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+func NewCircularExemplarStorage(len int, reg prometheus.Registerer) *CircularExemplarStorage {
+	return &CircularExemplarStorage{
+		exemplars: make([]*circularBufferEntry, len),
+		index:     make(map[string]int),
+		metrics:   newExemplarMetrics(reg),
+	}
+}
+
+func (ce *CircularExemplarStorage) Appender() storage.ExemplarAppender {
+	return ce
+}
+
+// TODO: separate wrapper struct for queries?
+func (ce *CircularExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (ce *CircularExemplarStorage) Select(start, end int64, l labels.Labels) ([]exemplar.ExemplarScrapeTimestamp, error) {
+
+	var (
+		ret []exemplar.ExemplarScrapeTimestamp
+		e   *circularBufferEntry
+		idx int
+		ok  bool
+		buf []byte
+	)
+
+	ce.lock.RLock()
+	defer ce.lock.RUnlock()
+
+	if idx, ok = ce.index[l.String()]; !ok {
+		return nil, nil
+	}
+	lastTs := ce.exemplars[idx].scrapeTimestamp
+
+	for {
+		// We need the labels check here in case what was the previous exemplar for the series
+		// when the exemplar from the last loop iteration was written has since been overwritten
+		// with an exemplar from another series.
+		// todo (callum) confirm if this check is still needed now that adding an exemplar should
+		// update the index and previous pointer for the series whose exemplar was overwritten.
+		if idx == -1 || string(ce.exemplars[idx].seriesLabels.Bytes(buf)) != string(l.Bytes(buf)) {
+			break
+		}
+
+		e = ce.exemplars[idx]
+		// todo (callum) This line is needed to avoid an infinite loop, consider redesign of buffer entry struct.
+		if ce.exemplars[idx].scrapeTimestamp > lastTs {
+			break
+		}
+
+		lastTs = ce.exemplars[idx].scrapeTimestamp
+		if e.scrapeTimestamp >= start && e.scrapeTimestamp <= end {
+			ret = append(ret, exemplar.ExemplarScrapeTimestamp{Exemplar: e.exemplar, ScrapeTimestamp: e.scrapeTimestamp})
+		}
+		idx = ce.exemplars[idx].prev
+	}
+	reverseExemplars(ret)
+	return ret, nil
+}
+
+// Takes the circularBufferEntry that will be overwritten and updates the
+// storages index for that entries labelset if necessary.
+func (ce *CircularExemplarStorage) indexGcCheck(cbe *circularBufferEntry) {
+	if cbe == nil {
+		return
+	}
+
+	l := cbe.seriesLabels
+	i := cbe.prev
+	if cbe.prev == -1 {
+		delete(ce.index, l.String())
+		return
+	}
+
+	l2 := ce.exemplars[i].seriesLabels
+	if !labels.Equal(l2, l) { // No more exemplars for series l.
```

Comment:
How can this ever be non-Equal? Surely it's a bug if the previous link is for a different series?

This should be dead code.